AsyncPing does not support ESP32

```console
~/github/trombik/esphome-component-ping/ esphome compile config/ping-esp32.yaml       
INFO Reading configuration config/ping-esp32.yaml...
INFO Generating C++ source...
INFO Core config or version changed, cleaning build files...
INFO Deleting config/ping_sensor/.pioenvs
INFO Deleting config/ping_sensor/.piolibdeps
INFO Compiling app...
INFO Running:  platformio run -d config/ping_sensor
Processing ping_sensor (board: lolin_d32; framework: arduino; platform: platformio/espressif32@3.2.0)
--------------------------------------------------------------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
PACKAGES: 
 - framework-arduinoespressif32 3.10006.210326 (1.0.6) 
 - tool-esptoolpy 1.30000.201119 (3.0.0) 
 - toolchain-xtensa32 2.50200.0 (5.2.0)
Library Manager: Installing esphome/AsyncTCP-esphome @ 1.2.2
Library Manager: AsyncTCP-esphome @ 1.2.2 has been installed!
Library Manager: Installing git+https://github.com/akaJes/AsyncPing#95ac7e4
git version 2.31.1
Cloning into '/usr/home/trombik/.platformio/.cache/tmp/pkg-installing-dsolfjai'...
HEAD is now at 95ac7e4 Fixes crash that occurs when an AsynPing object is destroyed while timers are running (#4)
Library Manager: AsyncPing(esp8266) @ 1.1.0+sha.95ac7e4 has been installed!
Library Manager: Installing ESPAsyncWebServer-esphome @ 1.2.7
Library Manager: ESPAsyncWebServer-esphome @ 1.2.7 has been installed!
Library Manager: Installing dependencies...
Library Manager: Installing ESPAsyncTCP-esphome
Library Manager: ESPAsyncTCP-esphome @ 1.2.3 has been installed!
Library Manager: Installing Hash
Library Manager: Already installed, built-in library
Library Manager: Installing Hash
Library Manager: Already installed, built-in library
Library Manager: Installing ArduinoJson-esphomelib @ 5.13.3
Library Manager: ArduinoJson-esphomelib @ 5.13.3 has been installed!
Dependency Graph
|-- <AsyncTCP-esphome> 1.2.2
|-- <AsyncPing(esp8266)> 1.1.0+sha.95ac7e4
|-- <FS> 1.0
|-- <ESPAsyncWebServer-esphome> 1.2.7
|   |-- <AsyncTCP-esphome> 1.2.2
|   |-- <FS> 1.0
|   |-- <WiFi> 1.0
|   |-- <ArduinoJson-esphomelib> 5.13.3
|-- <ESPmDNS> 1.0
|   |-- <WiFi> 1.0
|-- <ArduinoJson-esphomelib> 5.13.3
|-- <DNSServer> 1.1.0
|   |-- <WiFi> 1.0
|-- <Update> 1.0
|-- <WiFi> 1.0
Compiling .pioenvs/ping_sensor/src/esphome/components/captive_portal/captive_portal.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/components/json/json_util.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/components/logger/logger.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/components/ota/ota_component.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/components/sensor/automation.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/components/sensor/filter.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/components/sensor/sensor.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/components/web_server/web_server.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/components/web_server_base/web_server_base.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/components/wifi/wifi_component.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/components/wifi/wifi_component_esp32.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/components/wifi/wifi_component_esp8266.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/core/application.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/core/component.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/core/controller.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/core/esphal.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/core/helpers.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/core/log.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/core/preferences.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/core/scheduler.cpp.o
Compiling .pioenvs/ping_sensor/src/esphome/core/util.cpp.o
Compiling .pioenvs/ping_sensor/src/main.cpp.o
Generating partitions .pioenvs/ping_sensor/partitions.bin
In file included from src/esphome/components/ping/ping.h:3:0,
                 from src/esphome.h:6,
                 from src/main.cpp:3:
.piolibdeps/ping_sensor/AsyncPing_esp8266_/src/AsyncPing.h:8:30: fatal error: user_interface.h: No such file or directory
```

`esphome/components/wifi/wifi_component.h` has `ifdef`.

```cpp
#ifdef ARDUINO_ARCH_ESP32
#include <esp_wifi.h>
#include <WiFiType.h>
#include <WiFi.h>
#endif

#ifdef ARDUINO_ARCH_ESP8266
#include <ESP8266WiFiType.h>
#include <ESP8266WiFi.h>

#ifdef ARDUINO_ESP8266_RELEASE_2_3_0
extern "C" {
#include <user_interface.h>
};
#endif
#endif
```

--- a/.clang-format-ignore
+++ b/.clang-format-ignore
@@ -1,0 +1,3 @@
+# imported from esp-idf
+components/ping/ping_sock.c
+components/ping/ping_sock.h

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -6,6 +6,11 @@ on:
 jobs:
   build:
     runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        config:
+          - config/ping-esp32.yaml
+          - config/ping.yaml
     steps:
       - name: Checkout
         uses: actions/checkout@v2
@@ -26,4 +31,4 @@ jobs:
       - name: Compile ping.yaml
         run: |
           cp config/secrets.yaml.dist config/secrets.yaml
-          pipenv run esphome compile config/ping.yaml
+          pipenv run esphome compile ${{ matrix.config }}

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -40,3 +40,4 @@ jobs:
         with:
           clang-format-version: '11'
           check-path: 'components'
+          exclude-regex: components/ping/ping_sock.[ch]

--- a/README.md
+++ b/README.md
@@ -50,6 +50,8 @@ See [config/ping.yaml](config/ping.yaml).
 
 ## Log
 
+### `esp8266`
+
 ```console
 INFO Reading configuration config/ping.yaml...
 INFO Starting log output from ping.local using esphome API
@@ -95,4 +97,40 @@ INFO Successfully connected to ping.local
 [18:31:31][I][ping:037]: 64 bytes from 156.154.100.3: icmp_seq=12 ttl=55 time=43 ms
 [18:31:32][I][ping:037]: 64 bytes from 156.154.100.3: icmp_seq=13 ttl=55 time=41 ms
 [18:31:33][I][ping:063]: packet loss: 0.0 % latency: 40 ms
+```
+
+### `esp32`
+
+```console
+[C][logger:189]: Logger:
+[C][logger:190]:   Level: DEBUG
+[C][logger:191]:   Log Baud Rate: 115200
+[C][logger:192]:   Hardware UART: UART0
+[C][captive_portal:148]: Captive Portal:
+[C][web_server:136]: Web Server:
+[C][web_server:137]:   Address: ping_sensor.local:80
+[C][ota:029]: Over-The-Air Updates:
+[C][ota:030]:   Address: ping_sensor.local:3232
+[C][ota:032]:   Using Password.
+[I][ping_esp32:063]: 64 bytes from 8.8.8.8 icmp_seq=1 ttl=117 time=448 ms
+[I][ping_esp32:063]: 64 bytes from 8.8.8.8 icmp_seq=2 ttl=117 time=42 ms
+[I][ping_esp32:063]: 64 bytes from 8.8.8.8 icmp_seq=3 ttl=117 time=42 ms
+[I][ping_esp32:063]: 64 bytes from 8.8.8.8 icmp_seq=4 ttl=117 time=39 ms
+[I][ping_esp32:063]: 64 bytes from 8.8.8.8 icmp_seq=5 ttl=117 time=41 ms
+[I][ping_esp32:063]: 64 bytes from 8.8.8.8 icmp_seq=6 ttl=117 time=40 ms
+[I][ping_esp32:063]: 64 bytes from 8.8.8.8 icmp_seq=7 ttl=117 time=40 ms
+[I][ping_esp32:063]: 64 bytes from 8.8.8.8 icmp_seq=8 ttl=117 time=43 ms
+[I][ping_esp32:063]: 64 bytes from 8.8.8.8 icmp_seq=9 ttl=117 time=40 ms
+[I][ping_esp32:063]: 64 bytes from 8.8.8.8 icmp_seq=10 ttl=117 time=41 ms
+[I][ping_esp32:063]: 64 bytes from 8.8.8.8 icmp_seq=11 ttl=117 time=42 ms
+[I][ping_esp32:063]: 64 bytes from 8.8.8.8 icmp_seq=12 ttl=117 time=42 ms
+[I][ping_esp32:063]: 64 bytes from 8.8.8.8 icmp_seq=13 ttl=117 time=45 ms
+[I][ping_esp32:063]: 64 bytes from 8.8.8.8 icmp_seq=14 ttl=117 time=43 ms
+[I][ping_esp32:063]: 64 bytes from 8.8.8.8 icmp_seq=15 ttl=117 time=40 ms
+[I][ping_esp32:102]: From 8.8.8.8 icmp_seq=16 timeout
+[I][ping_esp32:102]: From 8.8.8.8 icmp_seq=17 timeout
+[I][ping_esp32:084]: --- 8.8.8.8 ping statistics ---
+[I][ping_esp32:089]: 17 packets transmitted, 15 received, 11% packet loss, total time 3028ms avg time 68ms
+[D][sensor:117]: 'Packet loss': Sending state 11.00000 % with 0 decimals of accuracy
+[D][sensor:117]: 'Latency': Sending state 0.06800 s with 3 decimals of accuracy
 ```

--- a/components/ping/ping.h
+++ b/components/ping/ping.h
@@ -1,16 +1,9 @@
 #pragma once
 
-#include "AsyncPing.h"
-
 #include "esphome/components/sensor/sensor.h"
-
-#define EACH_RESULT true
-#define END false
-#define TAG "ping"
 
 namespace esphome {
 namespace ping {
-
 class PingSensor : public sensor::Sensor, public PollingComponent {
  public:
   Sensor *packet_loss_sensor = new Sensor();
@@ -19,92 +12,28 @@ class PingSensor : public sensor::Sensor, public PollingComponent {
   /* the sensor must be started after connecting WiFi */
   float get_setup_priority() const override { return esphome::setup_priority::AFTER_WIFI; }
 
-  void setup() override {
-    ping.on(EACH_RESULT, [this](const AsyncPingResponse &response) {
-      if (response.answer) {
-        ESP_LOGI(TAG, "%d bytes from %s: icmp_seq=%d ttl=%d time=%d ms", response.size, target.c_str(),
-                 response.icmp_seq, response.ttl, response.time);
-
-        this->incr_total_success_time(response.time);
-      } else {
-        ESP_LOGI(TAG, "no reply from %s", target.c_str());
-      }
-      return false;
-    });
-
-    /* at the end, set the result */
-    ping.on(END, [this](const AsyncPingResponse &response) {
-      float loss = 0;
-      int total_failed_count = response.total_sent - response.total_recv;
-      if (response.total_sent != 0) {
-        loss = (float) total_failed_count / response.total_sent;
-      }
-
-      int mean = 0;
-      if (response.total_recv != 0) {
-        mean = total_success_time / response.total_recv;
-      }
-
-      this->set_latest_loss(loss * 100);
-      this->set_latest_latency(mean);
-
-      ESP_LOGI(TAG, "packet loss: %0.1f %% latency: %d ms", loss * 100, mean);
-      this->reset();
-      return true;
-    });
-    ping.begin(target.c_str(), n_packet, timeout);
-  }
-
-  void update() override {
-    float loss;
-    int latency_ms;
-
-    loss = this->get_latest_loss();
-    latency_ms = this->get_latest_latency();
-
-    if (loss >= 0 && this->packet_loss_sensor_ != nullptr) {
-      packet_loss_sensor_->publish_state(loss);
-    }
-    if (latency_ms >= 0 && this->latency_sensor_ != nullptr) {
-      latency_sensor_->publish_state((float) latency_ms / 1000);
-    }
-    ping.begin(target.c_str(), n_packet, timeout);
-  }
-
-  int get_latest_latency() { return latest_latency; }
-
-  float get_latest_loss() { return latest_loss; }
-
+  void set_n_packet(uint32_t n) { n_packet = n; }
+  void set_target(const std::string address) { target = address; }
   void set_timeout(uint32_t timeout_ms) { timeout = timeout_ms; }
 
-  void set_target(const std::string address) { target = address; }
-
-  void set_n_packet(uint32_t n) { n_packet = n; }
+  int get_latest_latency() { return latest_latency; }
+  float get_latest_loss() { return latest_loss; }
 
   void set_packet_loss_sensor(sensor::Sensor *packet_loss_sensor) { packet_loss_sensor_ = packet_loss_sensor; }
-
   void set_latency_sensor(sensor::Sensor *latency_sensor) { latency_sensor_ = latency_sensor; }
 
  private:
-  void set_latest_loss(float f) { latest_loss = f; }
-
-  void set_latest_latency(int i) { latest_latency = i; }
-
-  void reset() { total_success_time = 0; }
-
-  void incr_total_success_time(int time) { total_success_time += time; }
-
  protected:
-  float latest_loss = -1;
   int latest_latency = -1;
-  int total_success_time;
-  std::string target = "8.8.8.8";
   int timeout = 1000;
   int n_packet = 13;
-
-  AsyncPing ping;
+  float latest_loss = -1;
+  std::string target = "8.8.8.8";
   sensor::Sensor *packet_loss_sensor_{nullptr};
   sensor::Sensor *latency_sensor_{nullptr};
+
+  void set_latest_loss(float f) { latest_loss = f; }
+  void set_latest_latency(int i) { latest_latency = i; }
 };
 
 }  // namespace ping

--- a/components/ping/ping_esp32.h
+++ b/components/ping/ping_esp32.h
@@ -1,0 +1,141 @@
+#pragma once
+
+#ifdef ARDUINO_ARCH_ESP32
+#include "lwip/inet.h"
+#include "lwip/sockets.h"
+#include "ping_sock.h"
+
+#include "esphome/components/sensor/sensor.h"
+
+#define TAG "ping_esp32"
+
+namespace esphome {
+namespace ping {
+class PingSensorESP32 : public PingSensor {
+ public:
+  void setup() override {
+    init_ping();
+    esp_ping_new_session(&ping_config, &cbs, &ping);
+    esp_ping_start(ping);
+  }
+
+  void update() override {
+    float loss;
+    int latency_ms;
+
+    loss = this->get_latest_loss();
+    latency_ms = this->get_latest_latency();
+
+    if (loss >= 0 && this->packet_loss_sensor_ != nullptr) {
+      packet_loss_sensor_->publish_state(loss);
+    }
+    if (latency_ms >= 0 && this->latency_sensor_ != nullptr) {
+      latency_sensor_->publish_state((float) latency_ms / 1000);
+    }
+    esp_ping_new_session(&ping_config, &cbs, &ping);
+  }
+
+ private:
+  static void cb_ping_on_ping_success(esp_ping_handle_t hdl, void *context) {
+    reinterpret_cast<PingSensorESP32 *>(context)->cmd_ping_on_ping_success(hdl);
+  }
+
+  static void cb_cmd_ping_on_ping_end(esp_ping_handle_t hdl, void *context) {
+    reinterpret_cast<PingSensorESP32 *>(context)->cmd_ping_on_ping_end(hdl);
+  }
+
+  static void cb_cmd_ping_on_ping_timeout(esp_ping_handle_t hdl, void *context) {
+    reinterpret_cast<PingSensorESP32 *>(context)->cmd_ping_on_ping_timeout(hdl);
+  }
+
+  void cmd_ping_on_ping_success(esp_ping_handle_t hdl) {
+    uint8_t ttl;
+    uint16_t seqno;
+    uint32_t elapsed_time, recv_len;
+    ip_addr_t target_addr;
+    esp_ping_get_profile(hdl, ESP_PING_PROF_SEQNO, &seqno, sizeof(seqno));
+    esp_ping_get_profile(hdl, ESP_PING_PROF_TTL, &ttl, sizeof(ttl));
+    esp_ping_get_profile(hdl, ESP_PING_PROF_IPADDR, &target_addr, sizeof(target_addr));
+    esp_ping_get_profile(hdl, ESP_PING_PROF_SIZE, &recv_len, sizeof(recv_len));
+    esp_ping_get_profile(hdl, ESP_PING_PROF_TIMEGAP, &elapsed_time, sizeof(elapsed_time));
+    ESP_LOGI(TAG, "%d bytes from %s icmp_seq=%d ttl=%d time=%d ms", recv_len, ipaddr_ntoa((ip_addr_t *) &target_addr),
+             seqno, ttl, elapsed_time);
+    this->incr_total_success_time(elapsed_time);
+  }
+
+  void cmd_ping_on_ping_end(esp_ping_handle_t hdl) {
+    ip_addr_t target_addr;
+    uint32_t transmitted;
+    uint32_t received;
+    uint32_t total_time_ms;
+    int mean = 0;
+
+    esp_ping_get_profile(hdl, ESP_PING_PROF_REQUEST, &transmitted, sizeof(transmitted));
+    esp_ping_get_profile(hdl, ESP_PING_PROF_REPLY, &received, sizeof(received));
+    esp_ping_get_profile(hdl, ESP_PING_PROF_IPADDR, &target_addr, sizeof(target_addr));
+    esp_ping_get_profile(hdl, ESP_PING_PROF_DURATION, &total_time_ms, sizeof(total_time_ms));
+    uint32_t loss = (uint32_t)((1 - ((float) received) / transmitted) * 100);
+    if (received != 0) {
+      mean = total_success_time / received;
+    }
+
+    if (IP_IS_V4(&target_addr)) {
+      ESP_LOGI(TAG, "--- %s ping statistics ---", inet_ntoa(*ip_2_ip4(&target_addr)));
+    } else {
+      ESP_LOGI(TAG, "--- %s ping statistics ---", inet6_ntoa(*ip_2_ip6(&target_addr)));
+    }
+    ESP_LOGI(TAG, "%d packets transmitted, %d received, %d%% packet loss, total time %dms avg time %dms", transmitted,
+             received, loss, total_time_ms, mean);
+
+    this->set_latest_loss(loss);
+    this->set_latest_latency(mean);
+    this->reset();
+    esp_ping_delete_session(hdl);
+  }
+
+  void cmd_ping_on_ping_timeout(esp_ping_handle_t hdl) {
+    uint16_t seqno;
+    ip_addr_t target_addr;
+    esp_ping_get_profile(hdl, ESP_PING_PROF_SEQNO, &seqno, sizeof(seqno));
+    esp_ping_get_profile(hdl, ESP_PING_PROF_IPADDR, &target_addr, sizeof(target_addr));
+    ESP_LOGI(TAG, "From %s icmp_seq=%d timeout", ipaddr_ntoa((ip_addr_t *) &target_addr), seqno);
+  }
+
+ protected:
+  std::string tag = "ping_esp32";
+  esp_ping_config_t ping_config = ESP_PING_DEFAULT_CONFIG();
+  esp_ping_handle_t ping;
+  esp_ping_callbacks_t cbs;
+
+  void init_ping() {
+    ip_addr_t target_addr;
+    int err;
+
+    memset(&target_addr, 0, sizeof(target_addr));
+    err = inet_pton(AF_INET, target.c_str(), &target_addr);
+    if (err == 0) {
+      ESP_LOGE(tag.c_str(), "invalid address: `%s`", target.c_str());
+      this->status_set_warning();
+      return;
+    } else if (err < 0) {
+      ESP_LOGE(tag.c_str(), "inet_pton(): %s", esp_err_to_name(errno));
+      this->status_set_warning();
+      return;
+    }
+
+    ping_config.target_addr = target_addr;
+    ping_config.count = n_packet;
+
+    cbs.on_ping_success = PingSensorESP32::cb_ping_on_ping_success;
+    cbs.on_ping_timeout = PingSensorESP32::cb_cmd_ping_on_ping_timeout;
+    cbs.on_ping_end = PingSensorESP32::cb_cmd_ping_on_ping_end;
+    cbs.cb_args = this;
+  }
+  int total_success_time;
+  void reset() { total_success_time = 0; }
+  void incr_total_success_time(int time) { total_success_time += time; }
+};
+
+}  // namespace ping
+}  // namespace esphome
+#endif

--- a/components/ping/ping_esp8266.h
+++ b/components/ping/ping_esp8266.h
@@ -1,0 +1,77 @@
+#pragma once
+
+#include "AsyncPing.h"
+
+#include "esphome/components/sensor/sensor.h"
+
+#define EACH_RESULT true
+#define END false
+#define TAG "ping"
+
+namespace esphome {
+namespace ping {
+class PingSensorESP8266 : public PingSensor {
+ public:
+  void setup() override {
+    ping.on(EACH_RESULT, [this](const AsyncPingResponse &response) {
+      if (response.answer) {
+        ESP_LOGI(TAG, "%d bytes from %s: icmp_seq=%d ttl=%d time=%d ms", response.size, target.c_str(),
+                 response.icmp_seq, response.ttl, response.time);
+
+        this->incr_total_success_time(response.time);
+      } else {
+        ESP_LOGI(TAG, "no reply from %s", target.c_str());
+      }
+      return false;
+    });
+
+    /* at the end, set the result */
+    ping.on(END, [this](const AsyncPingResponse &response) {
+      float loss = 0;
+      int total_failed_count = response.total_sent - response.total_recv;
+      if (response.total_sent != 0) {
+        loss = (float) total_failed_count / response.total_sent;
+      }
+
+      int mean = 0;
+      if (response.total_recv != 0) {
+        mean = total_success_time / response.total_recv;
+      }
+
+      this->set_latest_loss(loss * 100);
+      this->set_latest_latency(mean);
+
+      ESP_LOGI(TAG, "packet loss: %0.1f %% latency: %d ms", loss * 100, mean);
+      this->reset();
+      return true;
+    });
+    ping.begin(target.c_str(), n_packet, timeout);
+  }
+
+  void update() override {
+    float loss;
+    int latency_ms;
+
+    loss = this->get_latest_loss();
+    latency_ms = this->get_latest_latency();
+
+    if (loss >= 0 && this->packet_loss_sensor_ != nullptr) {
+      packet_loss_sensor_->publish_state(loss);
+    }
+    if (latency_ms >= 0 && this->latency_sensor_ != nullptr) {
+      latency_sensor_->publish_state((float) latency_ms / 1000);
+    }
+    ping.begin(target.c_str(), n_packet, timeout);
+  }
+
+ private:
+ protected:
+  int total_success_time;
+  void reset() { total_success_time = 0; }
+  void incr_total_success_time(int time) { total_success_time += time; }
+
+  AsyncPing ping;
+};
+
+}  // namespace ping
+}  // namespace esphome

--- a/components/ping/ping_esp8266.h
+++ b/components/ping/ping_esp8266.h
@@ -1,5 +1,6 @@
 #pragma once
 
+#ifdef ARDUINO_ARCH_ESP8266
 #include "AsyncPing.h"
 
 #include "esphome/components/sensor/sensor.h"
@@ -75,3 +76,4 @@ class PingSensorESP8266 : public PingSensor {
 
 }  // namespace ping
 }  // namespace esphome
+#endif

--- a/components/ping/ping_sock.c
+++ b/components/ping/ping_sock.c
@@ -1,0 +1,409 @@
+// Copyright 2019 Espressif Systems (Shanghai) PTE LTD
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <stdlib.h>
+#include <stdbool.h>
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "lwip/opt.h"
+#include "lwip/init.h"
+#include "lwip/mem.h"
+#include "lwip/icmp.h"
+#include "lwip/netif.h"
+#include "lwip/sys.h"
+#include "lwip/timeouts.h"
+#include "lwip/inet.h"
+#include "lwip/inet_chksum.h"
+#include "lwip/ip.h"
+#include "lwip/netdb.h"
+#include "lwip/sockets.h"
+#include "esp_log.h"
+#include "ping/ping_sock.h"
+
+const static char *TAG = "ping_sock";
+
+#define PING_CHECK(a, str, goto_tag, ret_value, ...)                              \
+    do                                                                            \
+    {                                                                             \
+        if (!(a))                                                                 \
+        {                                                                         \
+            ESP_LOGE(TAG, "%s(%d): " str, __FUNCTION__, __LINE__, ##__VA_ARGS__); \
+            ret = ret_value;                                                      \
+            goto goto_tag;                                                        \
+        }                                                                         \
+    } while (0)
+
+#define PING_TIME_DIFF_MS(_end, _start) ((uint32_t)(((_end).tv_sec - (_start).tv_sec) * 1000 + \
+                                                    ((_end).tv_usec - (_start).tv_usec) / 1000))
+
+#define PING_CHECK_START_TIMEOUT_MS (1000)
+
+#define PING_FLAGS_INIT (1 << 0)
+#define PING_FLAGS_START (1 << 1)
+
+typedef struct {
+    int sock;
+    struct sockaddr_storage target_addr;
+    TaskHandle_t ping_task_hdl;
+    struct icmp_echo_hdr *packet_hdr;
+    ip_addr_t recv_addr;
+    uint32_t recv_len;
+    uint32_t icmp_pkt_size;
+    uint32_t count;
+    uint32_t transmitted;
+    uint32_t received;
+    uint32_t interval_ms;
+    uint32_t elapsed_time_ms;
+    uint32_t total_time_ms;
+    uint8_t ttl;
+    uint32_t flags;
+    void (*on_ping_success)(esp_ping_handle_t hdl, void *args);
+    void (*on_ping_timeout)(esp_ping_handle_t hdl, void *args);
+    void (*on_ping_end)(esp_ping_handle_t hdl, void *args);
+    void *cb_args;
+} esp_ping_t;
+
+static esp_err_t esp_ping_send(esp_ping_t *ep)
+{
+    esp_err_t ret = ESP_OK;
+    ep->packet_hdr->seqno++;
+    /* generate checksum since "seqno" has changed */
+    ep->packet_hdr->chksum = 0;
+    if (ep->packet_hdr->type == ICMP_ECHO) {
+        ep->packet_hdr->chksum = inet_chksum(ep->packet_hdr, ep->icmp_pkt_size);
+    }
+
+    ssize_t sent = sendto(ep->sock, ep->packet_hdr, ep->icmp_pkt_size, 0,
+                      (struct sockaddr *)&ep->target_addr, sizeof(ep->target_addr));
+
+    if (sent != (ssize_t)ep->icmp_pkt_size) {
+        int opt_val;
+        socklen_t opt_len = sizeof(opt_val);
+        getsockopt(ep->sock, SOL_SOCKET, SO_ERROR, &opt_val, &opt_len);
+        ESP_LOGE(TAG, "send error=%d", opt_val);
+        ret = ESP_FAIL;
+    } else {
+        ep->transmitted++;
+    }
+    return ret;
+}
+
+static int esp_ping_receive(esp_ping_t *ep)
+{
+    char buf[64]; // 64 bytes are enough to cover IP header and ICMP header
+    int len = 0;
+    struct sockaddr_storage from;
+    int fromlen = sizeof(from);
+    uint16_t data_head = 0;
+
+    while ((len = recvfrom(ep->sock, buf, sizeof(buf), 0, (struct sockaddr *)&from, (socklen_t *)&fromlen)) > 0) {
+        if (from.ss_family == AF_INET) {
+            // IPv4
+            struct sockaddr_in *from4 = (struct sockaddr_in *)&from;
+            inet_addr_to_ip4addr(ip_2_ip4(&ep->recv_addr), &from4->sin_addr);
+            IP_SET_TYPE_VAL(ep->recv_addr, IPADDR_TYPE_V4);
+            data_head = (uint16_t)(sizeof(struct ip_hdr) + sizeof(struct icmp_echo_hdr));
+        }
+#if CONFIG_LWIP_IPV6
+        else {
+            // IPv6
+            struct sockaddr_in6 *from6 = (struct sockaddr_in6 *)&from;
+            inet6_addr_to_ip6addr(ip_2_ip6(&ep->recv_addr), &from6->sin6_addr);
+            IP_SET_TYPE_VAL(ep->recv_addr, IPADDR_TYPE_V6);
+            data_head = (uint16_t)(sizeof(struct ip6_hdr) + sizeof(struct icmp6_echo_hdr));
+        }
+#endif
+        if (len >= data_head) {
+            if (IP_IS_V4_VAL(ep->recv_addr)) {              // Currently we process IPv4
+                struct ip_hdr *iphdr = (struct ip_hdr *)buf;
+                struct icmp_echo_hdr *iecho = (struct icmp_echo_hdr *)(buf + (IPH_HL(iphdr) * 4));
+                if ((iecho->id == ep->packet_hdr->id) && (iecho->seqno == ep->packet_hdr->seqno)) {
+                    ep->received++;
+                    ep->ttl = iphdr->_ttl;
+                    ep->recv_len = lwip_ntohs(IPH_LEN(iphdr)) - data_head;  // The data portion of ICMP
+                    return len;
+                }
+            }
+#if CONFIG_LWIP_IPV6
+            else if (IP_IS_V6_VAL(ep->recv_addr)) {      // Currently we process IPv6
+                struct ip6_hdr *iphdr = (struct ip6_hdr *)buf;
+                struct icmp6_echo_hdr *iecho6 = (struct icmp6_echo_hdr *)(buf + sizeof(struct ip6_hdr)); // IPv6 head length is 40
+                if ((iecho6->id == ep->packet_hdr->id) && (iecho6->seqno == ep->packet_hdr->seqno)) {
+                    ep->received++;
+                    ep->recv_len = IP6H_PLEN(iphdr) - sizeof(struct icmp6_echo_hdr); //The data portion of ICMPv6
+                    return len;
+                }
+            }
+#endif
+        }
+        fromlen = sizeof(from);
+    }
+    // if timeout, len will be -1
+    return len;
+}
+
+static void esp_ping_thread(void *args)
+{
+    esp_ping_t *ep = (esp_ping_t *)(args);
+    TickType_t last_wake;
+    struct timeval start_time, end_time;
+    int recv_ret;
+
+    while (1) {
+        /* wait for ping start signal */
+        if (ulTaskNotifyTake(pdTRUE, pdMS_TO_TICKS(PING_CHECK_START_TIMEOUT_MS))) {
+            /* initialize runtime statistics */
+            ep->packet_hdr->seqno = 0;
+            ep->transmitted = 0;
+            ep->received = 0;
+            ep->total_time_ms = 0;
+
+            last_wake = xTaskGetTickCount();
+            while ((ep->flags & PING_FLAGS_START) && ((ep->count == 0) || (ep->packet_hdr->seqno < ep->count))) {
+                esp_ping_send(ep);
+                gettimeofday(&start_time, NULL);
+                recv_ret = esp_ping_receive(ep);
+                gettimeofday(&end_time, NULL);
+                ep->elapsed_time_ms = PING_TIME_DIFF_MS(end_time, start_time);
+                ep->total_time_ms += ep->elapsed_time_ms;
+                if (recv_ret >= 0) {
+                    if (ep->on_ping_success) {
+                        ep->on_ping_success((esp_ping_handle_t)ep, ep->cb_args);
+                    }
+                } else {
+                    if (ep->on_ping_timeout) {
+                        ep->on_ping_timeout((esp_ping_handle_t)ep, ep->cb_args);
+                    }
+                }
+                if (pdMS_TO_TICKS(ep->interval_ms)) {
+                    vTaskDelayUntil(&last_wake, pdMS_TO_TICKS(ep->interval_ms)); // to get a more accurate delay
+                }
+            }
+            /* batch of ping operations finished */
+            if (ep->on_ping_end) {
+                ep->on_ping_end((esp_ping_handle_t)ep, ep->cb_args);
+            }
+        } else {
+            // check if ping has been de-initialized
+            if (!(ep->flags & PING_FLAGS_INIT)) {
+                break;
+            }
+        }
+    }
+    /* before exit task, free all resources */
+    if (ep->packet_hdr) {
+        free(ep->packet_hdr);
+    }
+    if (ep->sock > 0) {
+        close(ep->sock);
+    }
+    free(ep);
+    vTaskDelete(NULL);
+}
+
+esp_err_t esp_ping_new_session(const esp_ping_config_t *config, const esp_ping_callbacks_t *cbs, esp_ping_handle_t *hdl_out)
+{
+    esp_err_t ret = ESP_OK;
+    esp_ping_t *ep = NULL;
+    PING_CHECK(config, "ping config can't be null", err, ESP_ERR_INVALID_ARG);
+    PING_CHECK(hdl_out, "ping handle can't be null", err, ESP_ERR_INVALID_ARG);
+
+    ep = mem_calloc(1, sizeof(esp_ping_t));
+    PING_CHECK(ep, "no memory for esp_ping object", err, ESP_ERR_NO_MEM);
+
+    /* set INIT flag, so that ping task won't exit (must set before create ping task) */
+    ep->flags |= PING_FLAGS_INIT;
+
+    /* create ping thread */
+    BaseType_t xReturned = xTaskCreate(esp_ping_thread, "ping", config->task_stack_size, ep,
+                                       config->task_prio, &ep->ping_task_hdl);
+    PING_CHECK(xReturned == pdTRUE, "create ping task failed", err, ESP_ERR_NO_MEM);
+
+    /* callback functions */
+    if (cbs) {
+        ep->cb_args = cbs->cb_args;
+        ep->on_ping_end = cbs->on_ping_end;
+        ep->on_ping_timeout = cbs->on_ping_timeout;
+        ep->on_ping_success = cbs->on_ping_success;
+    }
+    /* set parameters for ping */
+    ep->recv_addr = config->target_addr;
+    ep->count = config->count;
+    ep->interval_ms = config->interval_ms;
+    ep->icmp_pkt_size = sizeof(struct icmp_echo_hdr) + config->data_size;
+    ep->packet_hdr = mem_calloc(1, ep->icmp_pkt_size);
+    PING_CHECK(ep->packet_hdr, "no memory for echo packet", err, ESP_ERR_NO_MEM);
+    /* set ICMP type and code field */
+    ep->packet_hdr->code = 0;
+    /* ping id should be unique, treat task handle as ping ID */
+    ep->packet_hdr->id = ((uint32_t)ep->ping_task_hdl) & 0xFFFF;
+    /* fill the additional data buffer with some data */
+    char *d = (char *)(ep->packet_hdr) + sizeof(struct icmp_echo_hdr);
+    for (uint32_t i = 0; i < config->data_size; i++) {
+        d[i] = 'A' + i;
+    }
+
+    /* create socket */
+    if (IP_IS_V4(&config->target_addr)
+#if CONFIG_LWIP_IPV6
+        || ip6_addr_isipv4mappedipv6(ip_2_ip6(&config->target_addr))
+#endif
+    ) {
+        ep->sock = socket(AF_INET, SOCK_RAW, IP_PROTO_ICMP);
+    }
+#if CONFIG_LWIP_IPV6
+    else {
+        ep->sock = socket(AF_INET6, SOCK_RAW, IP6_NEXTH_ICMP6);
+    }
+#endif
+    PING_CHECK(ep->sock > 0, "create socket failed: %d", err, ESP_FAIL, ep->sock);
+    /* set if index */
+    if(config->interface) {
+        struct ifreq iface;
+        if(netif_index_to_name(config->interface, iface.ifr_name) == NULL) {
+          goto err;
+        }
+        if(setsockopt(ep->sock, SOL_SOCKET, SO_BINDTODEVICE, &iface, sizeof(iface) !=0)) {
+          goto err;
+        }
+    }
+    struct timeval timeout;
+    timeout.tv_sec = config->timeout_ms / 1000;
+    timeout.tv_usec = (config->timeout_ms % 1000) * 1000;
+    /* set receive timeout */
+    setsockopt(ep->sock, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
+
+    /* set tos */
+    setsockopt(ep->sock, IPPROTO_IP, IP_TOS, &config->tos, sizeof(config->tos));
+
+    /* set socket address */
+    if (IP_IS_V4(&config->target_addr)) {
+        struct sockaddr_in *to4 = (struct sockaddr_in *)&ep->target_addr;
+        to4->sin_family = AF_INET;
+        inet_addr_from_ip4addr(&to4->sin_addr, ip_2_ip4(&config->target_addr));
+        ep->packet_hdr->type = ICMP_ECHO;
+    }
+#if CONFIG_LWIP_IPV6
+    if (IP_IS_V6(&config->target_addr)) {
+        struct sockaddr_in6 *to6 = (struct sockaddr_in6 *)&ep->target_addr;
+        to6->sin6_family = AF_INET6;
+        inet6_addr_from_ip6addr(&to6->sin6_addr, ip_2_ip6(&config->target_addr));
+        ep->packet_hdr->type = ICMP6_TYPE_EREQ;
+    }
+#endif
+    /* return ping handle to user */
+    *hdl_out = (esp_ping_handle_t)ep;
+    return ESP_OK;
+err:
+    if (ep) {
+        if (ep->sock > 0) {
+            close(ep->sock);
+        }
+        if (ep->packet_hdr) {
+            free(ep->packet_hdr);
+        }
+        if (ep->ping_task_hdl) {
+            vTaskDelete(ep->ping_task_hdl);
+        }
+        free(ep);
+    }
+    return ret;
+}
+
+esp_err_t esp_ping_delete_session(esp_ping_handle_t hdl)
+{
+    esp_err_t ret = ESP_OK;
+    esp_ping_t *ep = (esp_ping_t *)hdl;
+    PING_CHECK(ep, "ping handle can't be null", err, ESP_ERR_INVALID_ARG);
+    /* reset init flags, then ping task will exit */
+    ep->flags &= ~PING_FLAGS_INIT;
+    return ESP_OK;
+err:
+    return ret;
+}
+
+esp_err_t esp_ping_start(esp_ping_handle_t hdl)
+{
+    esp_err_t ret = ESP_OK;
+    esp_ping_t *ep = (esp_ping_t *)hdl;
+    PING_CHECK(ep, "ping handle can't be null", err, ESP_ERR_INVALID_ARG);
+    ep->flags |= PING_FLAGS_START;
+    xTaskNotifyGive(ep->ping_task_hdl);
+    return ESP_OK;
+err:
+    return ret;
+}
+
+esp_err_t esp_ping_stop(esp_ping_handle_t hdl)
+{
+    esp_err_t ret = ESP_OK;
+    esp_ping_t *ep = (esp_ping_t *)hdl;
+    PING_CHECK(ep, "ping handle can't be null", err, ESP_ERR_INVALID_ARG);
+    ep->flags &= ~PING_FLAGS_START;
+    return ESP_OK;
+err:
+    return ret;
+}
+
+esp_err_t esp_ping_get_profile(esp_ping_handle_t hdl, esp_ping_profile_t profile, void *data, uint32_t size)
+{
+    esp_err_t ret = ESP_OK;
+    esp_ping_t *ep = (esp_ping_t *)hdl;
+    const void *from = NULL;
+    uint32_t copy_size = 0;
+    PING_CHECK(ep, "ping handle can't be null", err, ESP_ERR_INVALID_ARG);
+    PING_CHECK(data, "profile data can't be null", err, ESP_ERR_INVALID_ARG);
+    switch (profile) {
+    case ESP_PING_PROF_SEQNO:
+        from = &ep->packet_hdr->seqno;
+        copy_size = sizeof(ep->packet_hdr->seqno);
+        break;
+    case ESP_PING_PROF_TTL:
+        from = &ep->ttl;
+        copy_size = sizeof(ep->ttl);
+        break;
+    case ESP_PING_PROF_REQUEST:
+        from = &ep->transmitted;
+        copy_size = sizeof(ep->transmitted);
+        break;
+    case ESP_PING_PROF_REPLY:
+        from = &ep->received;
+        copy_size = sizeof(ep->received);
+        break;
+    case ESP_PING_PROF_IPADDR:
+        from = &ep->recv_addr;
+        copy_size = sizeof(ep->recv_addr);
+        break;
+    case ESP_PING_PROF_SIZE:
+        from = &ep->recv_len;
+        copy_size = sizeof(ep->recv_len);
+        break;
+    case ESP_PING_PROF_TIMEGAP:
+        from = &ep->elapsed_time_ms;
+        copy_size = sizeof(ep->elapsed_time_ms);
+        break;
+    case ESP_PING_PROF_DURATION:
+        from = &ep->total_time_ms;
+        copy_size = sizeof(ep->total_time_ms);
+        break;
+    default:
+        PING_CHECK(false, "unknow profile: %d", err, ESP_ERR_INVALID_ARG, profile);
+        break;
+    }
+    PING_CHECK(size >= copy_size, "unmatched data size for profile %d", err, ESP_ERR_INVALID_SIZE, profile);
+    memcpy(data, from, copy_size);
+    return ESP_OK;
+err:
+    return ret;
+}

--- a/components/ping/ping_sock.c
+++ b/components/ping/ping_sock.c
@@ -12,6 +12,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#ifdef ARDUINO_ARCH_ESP32
+
+#include <string.h>
 #include <stdlib.h>
 #include <stdbool.h>
 #include "freertos/FreeRTOS.h"
@@ -29,7 +32,7 @@
 #include "lwip/netdb.h"
 #include "lwip/sockets.h"
 #include "esp_log.h"
-#include "ping/ping_sock.h"
+#include "ping_sock.h"
 
 const static char *TAG = "ping_sock";
 
@@ -268,6 +271,10 @@ esp_err_t esp_ping_new_session(const esp_ping_config_t *config, const esp_ping_c
     }
 #endif
     PING_CHECK(ep->sock > 0, "create socket failed: %d", err, ESP_FAIL, ep->sock);
+#if 0
+    /* disable code that does not build with esphome.
+     * ifreq does not exist in esp-idf 3.x, which esphome uses */
+
     /* set if index */
     if(config->interface) {
         struct ifreq iface;
@@ -278,6 +285,7 @@ esp_err_t esp_ping_new_session(const esp_ping_config_t *config, const esp_ping_c
           goto err;
         }
     }
+#endif
     struct timeval timeout;
     timeout.tv_sec = config->timeout_ms / 1000;
     timeout.tv_usec = (config->timeout_ms % 1000) * 1000;
@@ -407,3 +415,4 @@ esp_err_t esp_ping_get_profile(esp_ping_handle_t hdl, esp_ping_profile_t profile
 err:
     return ret;
 }
+#endif

--- a/components/ping/ping_sock.h
+++ b/components/ping/ping_sock.h
@@ -1,0 +1,171 @@
+// Copyright 2019 Espressif Systems (Shanghai) PTE LTD
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+#include "esp_err.h"
+#include "lwip/ip_addr.h"
+
+/**
+* @brief Type of "ping" session handle
+*
+*/
+typedef void *esp_ping_handle_t;
+
+/**
+* @brief Type of "ping" callback functions
+*
+*/
+typedef struct {
+    /**
+    * @brief arguments for callback functions
+    *
+    */
+    void *cb_args;
+
+    /**
+    * @brief Invoked by internal ping thread when received ICMP echo reply packet
+    *
+    */
+    void (*on_ping_success)(esp_ping_handle_t hdl, void *args);
+
+    /**
+    * @brief Invoked by internal ping thread when receive ICMP echo reply packet timeout
+    *
+    */
+    void (*on_ping_timeout)(esp_ping_handle_t hdl, void *args);
+
+    /**
+    * @brief Invoked by internal ping thread when a ping session is finished
+    *
+    */
+    void (*on_ping_end)(esp_ping_handle_t hdl, void *args);
+} esp_ping_callbacks_t;
+
+/**
+* @brief Type of "ping" configuration
+*
+*/
+typedef struct {
+    uint32_t count;           /*!< A "ping" session contains count procedures */
+    uint32_t interval_ms;     /*!< Milliseconds between each ping procedure */
+    uint32_t timeout_ms;      /*!< Timeout value (in milliseconds) of each ping procedure */
+    uint32_t data_size;       /*!< Size of the data next to ICMP packet header */
+    uint8_t tos;              /*!< Type of Service, a field specified in the IP header */
+    ip_addr_t target_addr;    /*!< Target IP address, either IPv4 or IPv6 */
+    uint32_t task_stack_size; /*!< Stack size of internal ping task */
+    uint32_t task_prio;       /*!< Priority of internal ping task */
+    uint32_t interface;       /*!< Netif index, interface=0 means NETIF_NO_INDEX*/
+} esp_ping_config_t;
+
+/**
+ * @brief Default ping configuration
+ *
+ */
+#define ESP_PING_DEFAULT_CONFIG()        \
+    {                                    \
+        .count = 5,                      \
+        .interval_ms = 1000,             \
+        .timeout_ms = 1000,              \
+        .data_size = 64,                 \
+        .tos = 0,                        \
+        .target_addr = *(IP_ANY_TYPE),   \
+        .task_stack_size = 2048,         \
+        .task_prio = 2,                  \
+        .interface = 0,\
+    }
+
+#define ESP_PING_COUNT_INFINITE (0) /*!< Set ping count to zero will ping target infinitely */
+
+/**
+* @brief Profile of ping session
+*
+*/
+typedef enum {
+    ESP_PING_PROF_SEQNO,   /*!< Sequence number of a ping procedure */
+    ESP_PING_PROF_TTL,     /*!< Time to live of a ping procedure */
+    ESP_PING_PROF_REQUEST, /*!< Number of request packets sent out */
+    ESP_PING_PROF_REPLY,   /*!< Number of reply packets received */
+    ESP_PING_PROF_IPADDR,  /*!< IP address of replied target */
+    ESP_PING_PROF_SIZE,    /*!< Size of received packet */
+    ESP_PING_PROF_TIMEGAP, /*!< Elapsed time between request and reply packet */
+    ESP_PING_PROF_DURATION /*!< Elapsed time of the whole ping session */
+} esp_ping_profile_t;
+
+/**
+ * @brief Create a ping session
+ *
+ * @param config ping configuration
+ * @param cbs a bunch of callback functions invoked by internal ping task
+ * @param hdl_out handle of ping session
+ * @return
+ *      - ESP_ERR_INVALID_ARG: invalid parameters (e.g. configuration is null, etc)
+ *      - ESP_ERR_NO_MEM: out of memory
+ *      - ESP_FAIL: other internal error (e.g. socket error)
+ *      - ESP_OK: create ping session successfully, user can take the ping handle to do follow-on jobs
+ */
+esp_err_t esp_ping_new_session(const esp_ping_config_t *config, const esp_ping_callbacks_t *cbs, esp_ping_handle_t *hdl_out);
+
+/**
+ * @brief Delete a ping session
+ *
+ * @param hdl handle of ping session
+ * @return
+ *      - ESP_ERR_INVALID_ARG: invalid parameters (e.g. ping handle is null, etc)
+ *      - ESP_OK: delete ping session successfully
+ */
+esp_err_t esp_ping_delete_session(esp_ping_handle_t hdl);
+
+/**
+ * @brief Start the ping session
+ *
+ * @param hdl handle of ping session
+ * @return
+ *      - ESP_ERR_INVALID_ARG: invalid parameters (e.g. ping handle is null, etc)
+ *      - ESP_OK: start ping session successfully
+ */
+esp_err_t esp_ping_start(esp_ping_handle_t hdl);
+
+/**
+ * @brief Stop the ping session
+ *
+ * @param hdl handle of ping session
+ * @return
+ *      - ESP_ERR_INVALID_ARG: invalid parameters (e.g. ping handle is null, etc)
+ *      - ESP_OK: stop ping session successfully
+ */
+esp_err_t esp_ping_stop(esp_ping_handle_t hdl);
+
+/**
+ * @brief Get runtime profile of ping session
+ *
+ * @param hdl handle of ping session
+ * @param profile type of profile
+ * @param data profile data
+ * @param size profile data size
+ * @return
+ *      - ESP_ERR_INVALID_ARG: invalid parameters (e.g. ping handle is null, etc)
+ *      - ESP_ERR_INVALID_SIZE: the actual profile data size doesn't match the "size" parameter
+ *      - ESP_OK: get profile successfully
+ */
+esp_err_t esp_ping_get_profile(esp_ping_handle_t hdl, esp_ping_profile_t profile, void *data, uint32_t size);
+
+#ifdef __cplusplus
+}
+#endif

--- a/components/ping/ping_sock.h
+++ b/components/ping/ping_sock.h
@@ -12,6 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#ifdef ARDUINO_ARCH_ESP32
+
 #pragma once
 
 #ifdef __cplusplus
@@ -168,4 +170,6 @@ esp_err_t esp_ping_get_profile(esp_ping_handle_t hdl, esp_ping_profile_t profile
 
 #ifdef __cplusplus
 }
+#endif
+
 #endif

--- a/components/ping/sensor.py
+++ b/components/ping/sensor.py
@@ -17,7 +17,7 @@ DEPENDENCIES = ["network"]
 
 ping_ns = cg.esphome_ns.namespace("ping")
 
-PingSensor = ping_ns.class_("PingSensor", sensor.Sensor, cg.PollingComponent)
+PingSensor = ping_ns.class_("PingSensorESP8266", sensor.Sensor, cg.PollingComponent)
 
 
 def validate_timeout(n):

--- a/components/ping/sensor.py
+++ b/components/ping/sensor.py
@@ -12,13 +12,18 @@ from esphome.const import (
     ICON_EMPTY,
     DEVICE_CLASS_EMPTY,
 )
+from esphome.core import CORE
 
 DEPENDENCIES = ["network"]
 
 ping_ns = cg.esphome_ns.namespace("ping")
 
-PingSensor = ping_ns.class_("PingSensorESP8266", sensor.Sensor, cg.PollingComponent)
-
+if CORE.is_esp8266:
+    PingSensor = ping_ns.class_("PingSensorESP8266", sensor.Sensor, cg.PollingComponent)
+elif CORE.is_esp32:
+    PingSensor = ping_ns.class_("PingSensorESP32", sensor.Sensor, cg.PollingComponent)
+else:
+    raise NotImplementedError
 
 def validate_timeout(n):
     n = cv.positive_time_period_microseconds(n)

--- a/components/ping/sensor.py
+++ b/components/ping/sensor.py
@@ -25,6 +25,7 @@ elif CORE.is_esp32:
 else:
     raise NotImplementedError
 
+
 def validate_timeout(n):
     n = cv.positive_time_period_microseconds(n)
     if n.total_seconds > 8:

--- a/config/ping-esp32.yaml
+++ b/config/ping-esp32.yaml
@@ -1,0 +1,39 @@
+substitutions:
+  my_name: ping_sensor
+
+esphome:
+  name: ${my_name}
+  platform: ESP32
+  board: lolin32
+
+logger:
+
+web_server:
+
+ota:
+  password: !secret ota_password
+
+wifi:
+  ssid: !secret wifi_ssid
+  password: !secret wifi_password
+  ap:
+    ssid: Ping
+    password: !secret wifi_ap_password
+
+captive_portal:
+
+external_components:
+  - source:
+      type: local
+      path: ../components
+
+sensor:
+  - platform: ping
+    ip_address: 8.8.8.8
+    num_attempts: 17
+    timeout: 1sec
+    loss:
+      name: Packet loss
+    latency:
+      name: Latency
+      accuracy_decimals: 3

